### PR TITLE
MLA optimizations V1

### PR DESCRIPTION
**Pull request by johanna-rock-tt**
### Ticket
Link to Github Issue

### Problem description
MLA uses DRAM interleaved in many places and HP for wkv_b2 matmul.

### What's changed
Changed to use L1 sharded/L1 interleaved in MLa, use DP(replicated) wkv_b2 matmul, cleaned up.

### Checklist

- [ ] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml?query=branch:jrock/ds-l1-sharding)
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:jrock/ds-l1-sharding)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:jrock/ds-l1-sharding)
- [ ] New/Existing tests provide coverage for changes

#### Model tests

If your changes cover model-related code, you should run tests corresponding to affected models and platforms (Single card, T3K, Galaxy). "Choose your pipeline" workflows facilitate running multiple kinds of tests in a single run. Each offers `models-mandatory` and `models-extended` presets.
The former includes a minimal set of tests, to be run always. The latter extends that with additional ones - use your best judgement in deciding which is the most appropriate for your PR.

- [ ] [(Single) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:jrock/ds-l1-sharding)
  - [ ] `models-mandatory` preset (runs:
      [Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) and
      [Frequent model and ttnn tests](https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml))
  - [ ] `models-extended` preset (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml)
      and [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(T3K) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:jrock/ds-l1-sharding)
  - [ ] `models-mandatory` preset (runs:
      [Unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml))
  - [ ] `models-extended` preset
      (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) and
      [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-model-perf-tests.yaml) tests)
  - [ ] other selection - specify runs

- [ ] [(Galaxy) Choose your pipeline](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:jrock/ds-l1-sharding)
  - [ ] `models-mandatory` preset (runs:
      [Quick tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml))
  - [ ] `models-extended` preset
      (runs:
      the mandatory tests, plus [Demo](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) and
      [Model perf](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-model-perf-tests.yaml) tests)

- [ ] [(Galaxy) DeepSeek tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-deepseek-tests.yaml)
- [ ] [(Galaxy) DeepSeek demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml)

### CODE REVIEW

File: models/demos/deepseek_v3/tt/moe_gate.py [hunk 1]
```diff
@@ -136,71 +136,140 @@ def model_config(
 
         if mode == "decode":
             memory_config = ttnn.L1_MEMORY_CONFIG
+
+            USERS_PER_ROW = 32
```
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/deepseek_v3/tt/moe_gate.py]: done

File: models/demos/deepseek_v3/tt/moe_gate.py [hunk 2]
```diff
@@ -136,71 +136,140 @@ def model_config(
...
+            HIDDEN_SIZE = hf_config.hidden_size
```
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/deepseek_v3/tt/moe_gate.py]: done

File: models/demos/deepseek_v3/tt/mla/mla1d.py [hunk 3]
```diff
@@ -438,8 +465,21 @@ def decode_model_config(
...
         )
+        kv_rope_permute_config = PermuteConfig(
+            dims=(0, 2, 1, 3),
+            memory_config=ttnn.L1_MEMORY_CONFIG,
+        )
         kv_rope_out_reshard_config = ReshardConfig(
-            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            memory_config=ttnn.L1_MEMORY_CONFIG,
+        )
+        kv_rope_slice_config = SliceConfig(
```
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/deepseek_v3/tt/mla/mla1d.py]: done

File: models/demos/deepseek_v3/tt/mla/mla1d.py [hunk 4]
```diff
@@ -570,20 +583,50 @@ def decode_model_config(
...
-            memory_config=ttnn.DRAM_MEMORY_CONFIG,
+            in_dim=2,
+            out_dim=1,
+            memory_config=ttnn.L1_MEMORY_CONFIG,
             topology=ttnn.Topology.Linear,
         )
 
-        # FlashMLA all-to-all
+        wq_a2a_reshard_out_mem_config = ttnn.create_sharded_memory_config(
+            shape=(USERS_PER_ROW, 128, 576), core_grid=ttnn.CoreGrid(y=8, x=8), strategy=ttnn.ShardStrategy.HEIGHT
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/mla/mla1d.py]: Nit: Magic numbers

File: models/demos/deepseek_v3/tt/mla/mla1d.py [hunk 5]
```diff
@@ -800,65 +849,61 @@ def forward_decode(
         bsz = x.shape[2]
         scale = 1.0 / mla_tp_factor
 
-        # wq_a and wq_b
-        tt_q = ttnn.linear(x, **cfg["wq_a"])
-        tt_q = ttnn.experimental.reduce_scatter_minimal_async(
-            tt_q, **ccl.populate_reduce_scatter_runtime_args(cfg["wq_a_rs_decode"])
+        ##############################################
+        ### Fused Linear + AR: wq_kv_a (wq_a + wkv_a) ###
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/mla/mla1d.py]: Instead of a noisy banner we could make a helper function? I see there are a few places with these

File: models/demos/deepseek_v3/tt/mla/mla1d.py [hunk 6]
```diff
@@ -800,65 +849,61 @@ def forward_decode(
...
+        # 1,1,32,512 8x2 [32,32]
         tt_kv_nope = RMSNorm.forward_decode(tt_kv_nope, cfg["kv_norm"])
+        # 1,1,32,512 8x2 [32,32]
+        tt_kv_nope = ttnn.to_memory_config(tt_kv_nope, memory_config=ttnn.L1_MEMORY_CONFIG)
+        # 1,1,32,512 L1 interleaved
 
         # KV RoPE
-        tt_kv_rope = ttnn.permute(tt_kv_rope, (0, 2, 1, 3))  # [1, bsz, 1, qk_rope_head_dim]
-        tt_kv_rope = ttnn.to_memory_config(tt_kv_rope, **cfg["kv_rope_reshard"])
+        # 1,1,32,64 1x2 [32,32]
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/mla/mla1d.py:1145]: Would it make sense to use asserts on the expected shapes instead of comments which can become out of date and be incorrect?

File: models/demos/deepseek_v3/tt/generator.py [hunk 7]
```diff
@@ -522,6 +537,8 @@ def generate(
         # Run one or more prefill+decode batches
         for _ in range(repeat_batches):
             # Prefill
+            if self.signpost:
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/generator.py:569]: We could avoid these conditionals if we made `signpost = (header) -> None ` when not using tracy.

File: models/demos/deepseek_v3/tt/moe.py [hunk 8]
```diff
@@ -133,48 +133,98 @@ def model_config(
...
+                    #     strategy=ttnn.ShardStrategy.WIDTH,
+                    # ),
+                    memory_config=memory_config,
+                    cluster_axis=1,
+                    topology=ttnn.Topology.Linear,
+                ),
+            }
+        else:
+            memory_config = ttnn.DRAM_MEMORY_CONFIG
+            # Construct the config
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/moe.py:193]:
```suggestion
```

File: models/demos/deepseek_v3/tt/moe.py [hunk 9]
```diff
@@ -190,70 +240,107 @@ def prefill_model_config(
...
+        ###########
+        ### MOE ###
+        ###########
+
         x_rm = ttnn.to_layout(x, ttnn.ROW_MAJOR_LAYOUT)
-        x_rm = ttnn.reshape(x_rm, shape=(batch_size_per_device, 1, seq_len, cfg["hidden_size"]))
+        x_rm = ttnn.reshape(
+            x_rm,
+            shape=(batch_size_per_device, 1, seq_len, cfg["hidden_size"]),
+            # memory_config=cfg["x_rm_reshape_memory_config"],
```
Comment (esmalTT) [inline models/demos/deepseek_v3/tt/moe.py]: Maybe delete? Commented code could be confusing.

**Review by esmalTT — COMMENTED** [review]
A few small comments. Feel free to ignore any if they are too big of a headache to fix now, we can address them later.

File: models/demos/deepseek_v3/tt/moe.py [hunk 10]
```diff
@@ -133,48 +133,98 @@ def model_config(
...
+                    topology=ttnn.Topology.Linear,
+                ),
+                "revert_tp": AllGatherAsyncConfig(
+                    mesh_device=MeshDeviceStub(mesh_device.shape),
+                    dim=-1,  # Last dimension
+                    memory_config=ttnn.DRAM_MEMORY_CONFIG,
+                    cluster_axis=1,
+                    topology=ttnn.Topology.Linear,
+                ),
+            }
```
[the remark this answers is not in the document]
Comment (johanna-rock-tt) [reply models/demos/deepseek_v3/tt/moe.py:227]: Will keep separate for easier readability and changing decode/prefill independently.

File: models/demos/deepseek_v3/tt/mla/mla1d.py [hunk 11]
```diff
@@ -1088,13 +1055,288 @@ def forward_prefill(
...
+        # 1,1,32,2112 (q_lora_rank + kv_lora_rank + qk_rope_head_dim = 1536 + 512 + 64)
+
+        # AR using AG + local reduce (since sub-tile RS not supported for new shapes)
+        tt_q_kv = ttnn.experimental.all_gather_async(
+            tt_q_kv, **ccl.populate_all_gather_runtime_args(cfg["wq_kv_a_ag_decode"])
+        )  # [1, num_devices, bsz, q_lora_rank + kv_lora_rank + qk_rope_head_dim]
+        tt_q_kv = ttnn.experimental.fast_reduce_nc(
+            tt_q_kv,
+            **cfg["wq_kv_a_r_decode"],
+        )  # [1, 1, bsz, q_lora_rank + kv_lora_rank + qk_rope_head_dim]
```
Comment (kpaigwar) [inline models/demos/deepseek_v3/tt/mla/mla1d.py:1107]: Can AG+local_reduce-> all_reduce ?

File: models/demos/deepseek_v3/tt/moe.py [hunk 12]
```diff
@@ -190,70 +240,107 @@ def prefill_model_config(
 
     @classmethod
     def forward(cls, x: ttnn.Tensor, cfg: RunDecodeConfig | RunPrefillConfig) -> ttnn.Tensor:
-        # CCL runtime initialization in execution order
-        ccl = cfg["ccl"]
-
-        x = ttnn.experimental.all_gather_async(x, **ccl.populate_all_gather_runtime_args(cfg["revert_tp"]))
-
+        # breakpoint()
```
[the remark this answers is not in the document]
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py:243]: Please remove breakpoint

**Review by kpaigwar — APPROVED** [review]
Please check single and dual galaxy CI before merging

**Comment by yalrawwashTT** [comment]
Single galaxy module tests are failing, because the cache files for module tests aren't created in `/CI_new` dir. You'll likely need to run all the module tests locally with permissions to write to `/CI_new`, then running again will pass on CI.

File: models/demos/deepseek_v3/tt/mla/mla1d.py (hunk 11 above)
Comment (johanna-rock-tt) [reply models/demos/deepseek_v3/tt/mla/mla1d.py:1107]: Maybe now that would work, before the fusion that was not possible because of subtile support missing I think. In any case, this will be reworked by the ops team entirely, so I don't think we need to do anything about it now.

**Comment by yieldthought** [comment]
Codex review:

The MLA decode refactor now permutes `v_out` before the `wo_ag_decode` all_gather without updating the gather dimension, so multi-column meshes will all_gather across the batch dimension instead of heads and produce incorrect outputs. That correctness issue makes the patch unreliable even if other paths work.

Full review comments:

- [P1] Gather heads on correct dimension after wkv_b2 permute — https://github.com/tenstorrent/tt-metal/blob/55f689eb061b0c652297993fda47f243733b26ea/models/demos/deepseek_v3/tt/mla/mla1d.py#L1297
  In decode, `_fwd_decode_wkv_b2` now permutes `v_out` to `[1, bsz, num_heads_local, v_head_dim]` before `_fwd_decode_ag_reshape`, but `wo_ag_decode` still gathers with `dim=1`, which is the batch dimension after the permute. On multi-column meshes (TP>1) this all_gather concatenates batches instead of heads, so the reshape to `(bsz, num_heads * v_head_dim)` will be wrong and logits will be corrupted. Move the permute back after `wo_ag_decode` (as before) or update the all_gather config to `dim=2` so the head dimension is gathered.

- [P3] Honor --signpost flag when emitting Tracy signposts — https://github.com/tenstorrent/tt-metal/blob/55f689eb061b0c652297993fda47f243733b26ea/models/demos/deepseek_v3/tt/generator.py#L575
  The new `signpost` flag is stored on the generator but the `signpost()` call in `generate()` runs unconditionally; when `--signpost` is not provided, signpost markers are still emitted and the CLI flag has no effect. This affects any run that expects signposts to be off (default) and can add tracing overhead. Gate the signpost calls on `self.signpost`.

- [P3] Align run_demo profile default with CLI choices — https://github.com/tenstorrent/tt-metal/blob/55f689eb061b0c652297993fda47f243733b26ea/models/demos/deepseek_v3/demo/demo.py#L249
  The CLI `--profile` option only allows `all|decode|prefill` and defaults to `all`, but `run_demo()` now defaults `profile` to "both". Programmatic callers that omit this argument pass an unsupported value to `DeepseekGenerator`, so the profiling branches never enable Tracy. Consider defaulting to "all" or mapping "both" to "all".

**Comment by johanna-rock-tt** [comment]
[P1] Gather heads on correct dimension after wkv_b2 permute: not true, I changed a couple of things incl. weights being replicated now, gather is fine as it is.

**Comment by johanna-rock-tt** [comment]
[P3] Honor --signpost flag when emitting Tracy signposts
[P3] Align run_demo profile default with CLI choices

Done
